docs/config.md
# Configuring the application

Configuration is via environment.

## Minimal `.env`

```dotenv title="Minimal .env"
NAME=tmpl-starlite-saqlalchemy
DB_URL=postgresql+asyncpg://postgres:mysecretpassword@localhost:5432/postgres1
OPENAPI_CONTACT_EMAIL=peter.github@proton.me
OPENAPI_CONTACT_NAME="Peter Schutt"
OPENAPI_TITLE="Template starlite-saqlalchemy Application"
OPENAPI_VERSION=1.0.0
REDIS_URL=redis://localhost:6379/0
```

## Local Development

Structured logs are nice when sending our logs through to some ingestion service, however, not so
nice for local development.

set `ENVIRONMENT=local` in your local `.env` file for a nicer local development experience
(we implement
[this structlog pattern](https://www.structlog.org/en/stable/logging-best-practices.html#pretty-printing-vs-structured-output)
for you!).

## Full `.env`

```dotenv title="Example .env"
--8<-- ".env.example"
```


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.20.1 (2022-12-17)
### Fix
* Ensures PluginConfig doesn't carry forward refs downstream. ([#176](https://github.com/topsport-com-au/starlite-saqlalchemy/issues/176)) ([`b37c262`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/b37c26204bc18eeaff67d626d5089738fae1abf7))

## v0.20.0 (2022-12-17)
### Feature
* Enable toggling of db/cache checks in run-app script. ([#173](https://github.com/topsport-com-au/starlite-saqlalchemy/issues/173)) ([`16b6380`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/16b63803799670bca8ef4f89a945b6c9a1bbe3b5))

### Documentation
* Update CHANGELOG.md ([`c73da31`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/c73da3188b5b6467e96d10add4bee50f9a6ef012))

## v0.19.0 (2022-12-16)
### Feature
* Adds `service.RepositoryService` type. ([`598fc51`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/598fc519e3edc39475e6485ce65fcffeeceaebce)

### Breaking
* Adds `service.RepositoryService` type. ([`598fc51`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/598fc519e3edc39475e6485ce65fcffeeceaebce)

## v0.18.0 (2022-12-14)
### Feature
* "local" environment ([`2a439e7`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/2a439e7b723276831deb66ea068537cddeb43619))

### Fix
* Msgspec json log rendering ([`818539d`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/818539d19cd60b54fe136ee7efd4f8f462cc7ff5))

### Breaking
* "local" environment ([`2a439e7`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/2a439e7b723276831deb66ea068537cddeb43619))

## v0.17.0 (2022-12-13)
### Feature
* Follow starlite into msgspec. ([#165](https://github.com/topsport-com-au/starlite-saqlalchemy/issues/165)) ([`2104626`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/2104626ff9016ba97675c0489d62f5106f1a9757))

### Breaking
* follow starlite into msgspec. ([#165](https://github.com/topsport-com-au/starlite-saqlalchemy/issues/165)) ([`2104626`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/2104626ff9016ba97675c0489d62f5106f1a9757))

## v0.16.0 (2022-12-08)
### Feature
* **repo:** Abstract method `filter_collection_by_kwargs()` ([#159](https://github.com/topsport-com-au/starlite-saqlalchemy/issues/159)) ([`15bf7a8`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/15bf7a814da21e105fc26b98a876f68fa0c998b4))

## v0.15.0 (2022-12-05)
### Feature
* **testing:** ControllerTest utility. ([#152](https://github.com/topsport-com-au/starlite-saqlalchemy/issues/152)) ([`4cc707b`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/4cc707bf25b4493b0477df173040f0fa4454fe1c))

## v0.14.2 (2022-12-03)
### Fix
* **dto:** Fix UnionType parsing as relationship. ([`4b93d17`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/4b93d17e732849179bb1c2d014d8177a93a3dbfc))

## v0.14.1 (2022-11-30)
### Fix
* Fix for usage of mock repo. ([`dddf1e2`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/dddf1e2e52defbb7632368c8442417fa26076e74))

### Documentation
* **changelog:** Add missing entry ([`e88a510`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/e88a5103c7107707a864a99e7eeef25a8a68daa8))

## v0.14.0 (2022-11-28)
### Breaking
* Overhaul of DTO pattern ([`e503447`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/e5034479b028ef791a4947b5df5a14cf248e935c))

## v0.13.1 (2022-11-24)
### Fix
* **testing:** Fix GenericMockRepository collections. ([#136](https://github.com/topsport-com-au/starlite-saqlalchemy/issues/136)) ([`246c103`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/246c10305c8436f47142cd6208ba4bcf85017f17))

## v0.13.0 (2022-11-21)
### Feature
* Simplify service and repo. ([#134](https://github.com/topsport-com-au/starlite-saqlalchemy/issues/134)) ([`ed3b59a`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/ed3b59aea64d82bdc75ef4fca65b93b6fcf45d4c))

### Breaking
* simplify service and repo. ([#134](https://github.com/topsport-com-au/starlite-saqlalchemy/issues/134)) ([`ed3b59a`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/ed3b59aea64d82bdc75ef4fca65b93b6fcf45d4c))

## v0.12.1 (2022-11-21)
### Fix
* **log:** Exclude compressed body from log output. ([`4d47df5`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/4d47df59a7a2b4968921a5204a912be045075a93))

### Documentation
* **readme:** Add coverage badge. ([`3c44375`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/3c44375c81f94cf8214906af25312f1d52547fd0))

## v0.12.0 (2022-11-17)
### Feature
* **worker:** Set queue name so saq keys unique to application ([`c56a12f`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/c56a12f9390c2f8185fa5831115cca79d65e3120))

## v0.11.0 (2022-11-17)
### Feature
* **dto:** Relationship support + test for forward refs ([`1bb3fb3`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/1bb3fb3ceaddfb440531d582ea1b852792983629))
* **logging:** Configurable dependency log levels. ([`a4fda6d`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/a4fda6dc5de64c8e4ca45ac42d96fa808677439d))

## v0.10.0 (2022-11-17)
### Feature
* **service:** __class_getitem__ sets `repository_type` on service. ([`4127f3a`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/4127f3aec85560b1f17eaf90518f94d83e0ef5fd))
* **repository:** __class_getitem__ sets `model_type` on repo. ([`d81e6c6`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/d81e6c6df972c5d8d919e8d9c398404b57acfd8f))

## v0.9.0 (2022-11-17)
### Feature
* **logging:** Improve dev logging experience ([`fc04a99`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/fc04a9914d8210e9729bfc3857bc555d4ccfa1d1))
* **plugin:** Set starlite debug ([`1a9e01d`](https://github.com/topsport-com-au/starlite-saqlalchemy/commit/1a9e01d6e5a9b91e81c417e7af4bb5ce98dc586e))

### Fix

src/starlite_saqlalchemy/settings.py
"""All configuration via environment.

Take note of the environment variable prefixes required for each
settings class, except `AppSettings`.
"""
from __future__ import annotations

# pylint: disable=missing-class-docstring
from typing import Literal

from pydantic import AnyUrl, BaseSettings, PostgresDsn, parse_obj_as
from starlite.utils.extractors import RequestExtractorField, ResponseExtractorField


# noinspection PyUnresolvedReferences
class AppSettings(BaseSettings):
    """Generic application settings.

    These settings are returned as json by the healthcheck endpoint, so
    do not include any sensitive values here, or if you do ensure to
    exclude them from serialization in the `Config` object.
    """

    class Config:
        case_sensitive = True
        env_file = ".env"

    BUILD_NUMBER: str = ""
    """Identifier for CI build."""
    DEBUG: bool = False
    """Run `Starlite` with `debug=True`."""
    ENVIRONMENT: str = "prod"
    """'dev', 'prod', etc."""
    NAME: str = "my-starlite-saqlalchemy-app"
    """Application name."""

    @property
    def slug(self) -> str:
        """Return a slugified name.

        Returns:
            `self.NAME`, all lowercase and hyphens instead of spaces.
        """
        return "-".join(s.lower() for s in self.NAME.split())


# noinspection PyUnresolvedReferences
class APISettings(BaseSettings):
    """API specific configuration."""

    class Config:
        case_sensitive = True
        env_file = ".env"
        env_prefix = "API_"

    CACHE_EXPIRATION: int = 60
    """Default cache key expiration in seconds."""
    DB_SESSION_DEPENDENCY_KEY: str = "db_session"
    """Parameter name for SQLAlchemy session dependency injection."""
    DEFAULT_PAGINATION_LIMIT: int = 100
    """Max records received for collection routes."""
    DTO_INFO_KEY: str = "dto"
    """Key used for DTO field config in SQLAlchemy info dict."""
    HEALTH_PATH: str = "/health"
    """Route that the health check is served under."""


class LogSettings(BaseSettings):
    """Logging config for the application."""

    class Config:
        case_sensitive = True
        env_file = ".env"
        env_prefix = "LOG_"

    # https://stackoverflow.com/a/1845097/6560549
    EXCLUDE_PATHS: str = r"\A(?!x)x"
    """Regex to exclude paths from logging."""
    HTTP_EVENT: str = "HTTP"
    """Log event name for logs from Starlite handlers."""
    INCLUDE_COMPRESSED_BODY: bool = False
    """Include 'body' of compressed responses in log output."""
    LEVEL: int = 20
    """Stdlib log levels. Only emit logs at this level, or higher."""
    OBFUSCATE_COOKIES: set[str] = {"session"}
    """Request cookie keys to obfuscate."""
    OBFUSCATE_HEADERS: set[str] = {"Authorization", "X-API-KEY"}
    """Request header keys to obfuscate."""
    JOB_FIELDS: list[str] = [
        "function",
        "kwargs",
        "key",
        "scheduled",
        "attempts",
        "completed",
        "queued",
        "started",
        "result",
        "error",
    ]
    """Attributes of the SAQ [`Job`](https://github.com/tobymao/saq/blob/master/saq/job.py)
    to be logged.
    """
    REQUEST_FIELDS: list[RequestExtractorField] = [
        "path",
        "method",
        "content_type",
        "headers",
        "cookies",
        "query",
        "path_params",
        "body",
    ]
    """Attributes of the [Request][starlite.connection.request.Request] to be logged."""
    RESPONSE_FIELDS: list[ResponseExtractorField] = [
        "status_code",
        "cookies",
        "headers",
        "body",
    ]
    """Attributes of the [Response][starlite.response.Response] to be logged."""
    WORKER_EVENT: str = "Worker"
    """Log event name for logs from SAQ worker."""
    SAQ_LEVEL: int = 30
    """Level to log SAQ logs."""
    SQLALCHEMY_LEVEL: int = 30
    """Level to log SAQ logs."""
    UVICORN_ACCESS_LEVEL: int = 30
    """Level to log uvicorn access logs."""
    UVICORN_ERROR_LEVEL: int = 30
    """Level to log uvicorn error logs."""


# noinspection PyUnresolvedReferences
class OpenAPISettings(BaseSettings):
    """Configures OpenAPI for the application."""

    class Config:
        case_sensitive = True
        env_file = ".env"
        env_prefix = "OPENAPI_"

    CONTACT_NAME: str = "Peter"
    """Name of contact on document."""
    CONTACT_EMAIL: str = "peter.github@proton.me"
    """Email for contact on document."""
    TITLE: str | None = "My Starlite-SAQAlchemy App"
    """Document title."""
    VERSION: str = "v1.0"
    """Document version."""


# noinspection PyUnresolvedReferences
class DatabaseSettings(BaseSettings):
    """Configures the database for the application."""

    class Config:
        case_sensitive = True
        env_file = ".env"
        env_prefix = "DB_"

    ECHO: bool = False
    """Enable SQLAlchemy engine logs."""
    ECHO_POOL: bool | Literal["debug"] = False
    """Enable SQLAlchemy connection pool logs."""
    POOL_DISABLE: bool = False
    """Disable SQLAlchemy pooling, same as setting pool to
    [`NullPool`][sqlalchemy.pool.NullPool].
    """
    POOL_MAX_OVERFLOW: int = 10
    """See [`max_overflow`][sqlalchemy.pool.QueuePool]."""
    POOL_SIZE: int = 5
    """See [`pool_size`][sqlalchemy.pool.QueuePool]."""
    POOL_TIMEOUT: int = 30
    """See [`timeout`][sqlalchemy.pool.QueuePool]."""
    URL: PostgresDsn = parse_obj_as(
        PostgresDsn, "postgresql+asyncpg://postgres:mysecretpassword@localhost:5432/postgres"
    )


# noinspection PyUnresolvedReferences
class RedisSettings(BaseSettings):
    """Redis settings for the application."""

    class Config:
        case_sensitive = True
        env_file = ".env"
        env_prefix = "REDIS_"

    URL: AnyUrl = parse_obj_as(AnyUrl, "redis://localhost:6379/0")
    """A Redis connection URL."""


# noinspection PyUnresolvedReferences
class SentrySettings(BaseSettings):
    """Configures sentry for the application."""

    class Config:
        case_sensitive = True
        env_file = ".env"
        env_prefix = "SENTRY_"

    DSN: str = ""
    """The sentry DSN. Set as empty string to disable sentry reporting."""
    TRACES_SAMPLE_RATE: float = 0.0001
    """% of requests traced by sentry, `0.0` means none, `1.0` means all."""


class ServerSettings(BaseSettings):
    """Server configurations."""

    class Config:
        case_sensitive = True
        env_file = ".env"
        env_prefix = "SERVER_"

    APP_LOC: str = "app.main:create_app"
    """Path to app executable, or factory."""
    APP_LOC_IS_FACTORY: bool = True
    """Indicate if APP_LOC points to an executable or factory."""
    HOST: str = "localhost"
    """Server network host."""
    KEEPALIVE: int = 65
    """Seconds to hold connections open (65 is > AWS lb idle timeout)."""
    PORT: int = 8000
    """Server port."""
    RELOAD: bool = False
    """Turn on hot reloading."""
    RELOAD_DIRS: list[str] = ["src/"]
    """Directories to watch for reloading."""


# `.parse_obj()` thing is a workaround for pyright and pydantic interplay, see:
# https://github.com/pydantic/pydantic/issues/3753#issuecomment-1087417884
api = APISettings.parse_obj({})
"""Api settings."""
app = AppSettings.parse_obj({})
"""App settings."""
db = DatabaseSettings.parse_obj({})
"""Database settings."""
log = LogSettings.parse_obj({})
"""Log settings."""
openapi = OpenAPISettings.parse_obj({})
"""Openapi settings."""
redis = RedisSettings.parse_obj({})
"""Redis settings."""
sentry = SentrySettings.parse_obj({})
"""Sentry settings."""
server = ServerSettings.parse_obj({})
"""Server settings."""
